### DISCUSSION TITLE
How to check if an element exists on the webpage using Cypress

### DISCUSSION THREAD

**Discussion by sravanthi1047** [cypress-io/cypress#29389] [Questions and Help]
Hi All,

I am beginner in Cypress. Can anyone please help me by answering my question, it may be simple to u guys but not for me. Please help.

My requirement is such that based on the existence of an UI element on the application, for eg, if the searched user yields results on the users grid , the test should go and update its values otherwise it should go and create a new user with the desired details.

To perform this search I have tried using the following commands, but nothing is working out. Moreover, I am expecting the error should not be thrown when the searched element is "not" found on the application, but Cypress is observed to be raising the error.

1. var len=cy.get('#Id1').get("table").**filter**("#Id2").length
Error: 
![image](https://github.com/cypress-io/cypress/assets/65785048/9084dcec-bd9e-400c-8cad-e04ea80aefd4)

2. var len=cy.get('#Id1 table").**find**("#Id2").length
Error:
![image](https://github.com/cypress-io/cypress/assets/65785048/24537b00-e0d2-4edf-9736-587348a06111)

Can anyone pl let me know how can i perform this simple verification of UI element presence using Cypress, without having any error being raised in case that optional UI element misses on the application. Any help is greatly appreciated. Thanks

**Comment by portyu9** [comment]
Hi @sravanthi1047 

The main issue is that this does not give you the number of matched elements:

```js
var len = cy.get(...).length
```

Cypress commands are asynchronous and return a Cypress chain, not the DOM result directly.

Also, if the element is optional, calling `cy.get('#Id2')` to check whether it exists will fail when it does not exist.

Query a parent that is guaranteed to exist, then inspect its DOM synchronously:

```js
cy.get('#Id1').then(($container) => {
  const $user = $container.find('#Id2')

  if ($user.length) {
    // User exists - update it.
    cy.wrap($user).click()
  } else {
    // User does not exist - create it.
  }
})
```

One important caveat: only do this after the search results are known to have finished rendering. If the grid is populated by an API request, the more reliable pattern is to `cy.intercept()` that request, trigger the search, `cy.wait()` for it, and then perform the DOM check. Otherwise the conditional can become flaky.

Please mark this answer as accepted if it helped, thank you.
